### PR TITLE
Generate documentation in rustc `rustc_index::newtype_index` macro

### PR DESCRIPTION
**Pull request by pierwill**
The macro now documents all generated items. Documentation notes possible panics and unsafety.

### CODE REVIEW

File: compiler/rustc_index/src/vec.rs [hunk 1]
```diff
@@ -130,6 +137,11 @@ macro_rules! newtype_index {
                 }
             }
 
+            /// Creates a new index from a given `u32`.
+            ///
+            /// # Panics
+            ///
+            /// Will panic if `value` exceeds `MAX_AS_U32`.
```
Comment (camelid) [inline compiler/rustc_index/src/vec.rs:144]: IIUC, `MAX` and `MAX_AS_U32` are the same value, just represented differently (one as a `u32` and one as a newtype index). So maybe both here and `from_usize` should say `MAX`?

File: compiler/rustc_index/src/vec.rs [hunk 2]
```diff
@@ -138,12 +150,18 @@ macro_rules! newtype_index {
                 }
             }
 
+            /// Creates a new index from a given `u32`.
+            ///
+            /// # Safety
+            ///
+            /// This function is unsafe because it does not check for integer overflows.
```
Comment (camelid) [inline compiler/rustc_index/src/vec.rs:157]: I don't think "integer overflow" is quite right here; integer overflow is not unsafe.

Comment (pierwill) [reply compiler/rustc_index/src/vec.rs:157]: Why *does* this have to be `unsafe`? 🤔

Comment (pierwill) [reply compiler/rustc_index/src/vec.rs:157]: Removing the `unsafe`:

```
    = note: initializing a layout restricted type's field with a value outside the valid range is undefined behavior
```

Comment (pierwill) [reply compiler/rustc_index/src/vec.rs:157]: I think I'll just copy that into the doc.

Comment (camelid) [reply compiler/rustc_index/src/vec.rs:157]: I think because of this:

https://github.com/rust-lang/rust/blob/220ed09b26177ca4b6ab525f403d251024389a41/compiler/rustc_index/src/vec.rs#L108

Constructing a value outside the valid range would be UB.

Comment (pierwill) [reply compiler/rustc_index/src/vec.rs]: Thanks @camelid!

File: compiler/rustc_index/src/vec.rs [hunk 3]
```diff
@@ -156,6 +176,8 @@ macro_rules! newtype_index {
             }
 
             /// Extracts the value of this index as a `usize`.
+            ///
+            /// This casts a `u32` to a `usize`.
```
Comment (camelid) [inline compiler/rustc_index/src/vec.rs:180]: nit: Why did you add this comment here but not also for `index`? I think it should either be in both places or neither.

**Review by camelid — COMMENTED** [review]
I'll let @jackh726 or someone else from T-compiler take it from here :)

File: compiler/rustc_index/src/vec.rs [hunk 4]
```diff
@@ -138,12 +150,20 @@ macro_rules! newtype_index {
...
+            ///
+            /// # Safety
+            ///
+            /// This function is unsafe because initializing a layout restricted type's
+            /// field with a value outside the valid range is undefined behavior.
+            ///
+            /// Prefer using `from_u32`.
             #[inline]
             $v const unsafe fn from_u32_unchecked(value: u32) -> Self {
                 Self { private: value }
```
Comment (jyn514) [inline compiler/rustc_index/src/vec.rs:163]: This can't be true, right? Just constructing the index can't cause UB, only using it in an invalid way afterwards. Could you look in this module for where it depends on the invariant that `value < MAX`?

File: compiler/rustc_index/src/vec.rs [hunk 5]
```diff
@@ -130,6 +137,11 @@ macro_rules! newtype_index {
...
             }
 
+            /// Creates a new index from a given `u32`.
+            ///
+            /// # Panics
+            ///
+            /// Will panic if `value` exceeds `MAX`.
             #[inline]
             $v const fn from_u32(value: u32) -> Self {
                 assert!(value <= $max);
```
Comment (jackh726) [inline compiler/rustc_index/src/vec.rs:147]: While you're here, can you add a `SAFETY: we just checked that `value <= max`

File: compiler/rustc_index/src/vec.rs (hunk 4 above)
Comment (jackh726) [reply compiler/rustc_index/src/vec.rs:163]: It is because of `rustc_layout_scalar_valid_range_end`.

I think better docs here would be:
```
# Safety

The provided value must be less than or equal to the max value for the newtype. Providing a value outside this range is undefined due to layout restrictions.
```

**Review by jackh726 — CHANGES_REQUESTED** [review]
r=me with comment changed

**Comment by jackh726** [comment]
@bors delegate+ rollup

Just r=jackh726 when ready

**Comment by pierwill** [comment]
r=jackh726
